In [2]:
import pandas as pd
import numpy as np

In [3]:
df=pd.read_csv('responded_users.csv')
df=df.drop(columns='Unnamed: 0')
df=df.set_index('auction_id')
df

experiment        date  hour  \
auction_id                                                          
0016d14a-ae18-4a02-a204-6ba53b52f2ed    exposed  2020-07-05     2   
008aafdf-deef-4482-8fec-d98e3da054da    exposed  2020-07-04    16   
00a1384a-5118-4d1b-925b-6cdada50318d    exposed  2020-07-06     8   
00b6fadb-10bd-49e3-a778-290da82f7a8d    control  2020-07-08     4   
00ebf4a8-060f-4b99-93ac-c62724399483    control  2020-07-03    15   
...                                         ...         ...   ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc    exposed  2020-07-05    21   
ffb176df-ecd2-45d3-b05f-05b173a093a7    exposed  2020-07-04     1   
ffb79718-6f25-4896-b6b3-e58b80a6e147    control  2020-07-09     7   
ffca1153-c182-4f32-9e90-2a6008417497    control  2020-07-10    16   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b    exposed  2020-07-03     4   

                                             device_make  platform_os  \
auction_id                                                              
0016d14a-ae18-4a02-a204-6ba53b52f2ed               E5823            6   
008aafdf-deef-4482-8fec-d98e3da054da  Generic Smartphone            6   
00a1384a-5118-4d1b-925b-6cdada50318d  Generic Smartphone            6   
00b6fadb-10bd-49e3-a778-290da82f7a8d    Samsung SM-A202F            6   
00ebf4a8-060f-4b99-93ac-c62724399483  Generic Smartphone            6   
...                                                  ...          ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc  Generic Smartphone            6   
ffb176df-ecd2-45d3-b05f-05b173a093a7  Generic Smartphone            6   
ffb79718-6f25-4896-b6b3-e58b80a6e147  Generic Smartphone            6   
ffca1153-c182-4f32-9e90-2a6008417497  Generic Smartphone            6   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b  Generic Smartphone            6   

                                                    browser  yes  no  \
auction_id                                                             
0016d14a-ae18-4a02-a204-6ba53b52f2ed  Chrome Mobile WebView    0   1   
008aafdf-deef-4482-8fec-d98e3da054da          Chrome Mobile    1   0   
00a1384a-5118-4d1b-925b-6cdada50318d          Chrome Mobile    0   1   
00b6fadb-10bd-49e3-a778-290da82f7a8d               Facebook    1   0   
00ebf4a8-060f-4b99-93ac-c62724399483          Chrome Mobile    0   1   
...                                                     ...  ...  ..   
ffa08ff9-a132-4051-aef5-01a9c79367bc          Chrome Mobile    1   0   
ffb176df-ecd2-45d3-b05f-05b173a093a7          Chrome Mobile    1   0   
ffb79718-6f25-4896-b6b3-e58b80a6e147          Chrome Mobile    0   1   
ffca1153-c182-4f32-9e90-2a6008417497          Chrome Mobile    0   1   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b          Chrome Mobile    0   1   

                                      response day_of_the_week  
auction_id                                                      
0016d14a-ae18-4a02-a204-6ba53b52f2ed       1.0          Sunday  
008aafdf-deef-4482-8fec-d98e3da054da       1.0        Saturday  
00a1384a-5118-4d1b-925b-6cdada50318d       1.0          Monday  
00b6fadb-10bd-49e3-a778-290da82f7a8d       1.0       Wednesday  
00ebf4a8-060f-4b99-93ac-c62724399483       1.0          Friday  
...                                        ...             ...  
ffa08ff9-a132-4051-aef5-01a9c79367bc       1.0          Sunday  
ffb176df-ecd2-45d3-b05f-05b173a093a7       1.0        Saturday  
ffb79718-6f25-4896-b6b3-e58b80a6e147       1.0        Thursday  
ffca1153-c182-4f32-9e90-2a6008417497       1.0          Friday  
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b       1.0          Friday  

[1243 rows x 10 columns]

In [4]:
#users who responded yes rep by 1 and those who responded no by 0
df.loc[(df['yes']==1),'responses']=1
df['responses']=df['responses'].fillna(0)
df

experiment        date  hour  \
auction_id                                                          
0016d14a-ae18-4a02-a204-6ba53b52f2ed    exposed  2020-07-05     2   
008aafdf-deef-4482-8fec-d98e3da054da    exposed  2020-07-04    16   
00a1384a-5118-4d1b-925b-6cdada50318d    exposed  2020-07-06     8   
00b6fadb-10bd-49e3-a778-290da82f7a8d    control  2020-07-08     4   
00ebf4a8-060f-4b99-93ac-c62724399483    control  2020-07-03    15   
...                                         ...         ...   ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc    exposed  2020-07-05    21   
ffb176df-ecd2-45d3-b05f-05b173a093a7    exposed  2020-07-04     1   
ffb79718-6f25-4896-b6b3-e58b80a6e147    control  2020-07-09     7   
ffca1153-c182-4f32-9e90-2a6008417497    control  2020-07-10    16   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b    exposed  2020-07-03     4   

                                             device_make  platform_os  \
auction_id                                                              
0016d14a-ae18-4a02-a204-6ba53b52f2ed               E5823            6   
008aafdf-deef-4482-8fec-d98e3da054da  Generic Smartphone            6   
00a1384a-5118-4d1b-925b-6cdada50318d  Generic Smartphone            6   
00b6fadb-10bd-49e3-a778-290da82f7a8d    Samsung SM-A202F            6   
00ebf4a8-060f-4b99-93ac-c62724399483  Generic Smartphone            6   
...                                                  ...          ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc  Generic Smartphone            6   
ffb176df-ecd2-45d3-b05f-05b173a093a7  Generic Smartphone            6   
ffb79718-6f25-4896-b6b3-e58b80a6e147  Generic Smartphone            6   
ffca1153-c182-4f32-9e90-2a6008417497  Generic Smartphone            6   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b  Generic Smartphone            6   

                                                    browser  yes  no  \
auction_id                                                             
0016d14a-ae18-4a02-a204-6ba53b52f2ed  Chrome Mobile WebView    0   1   
008aafdf-deef-4482-8fec-d98e3da054da          Chrome Mobile    1   0   
00a1384a-5118-4d1b-925b-6cdada50318d          Chrome Mobile    0   1   
00b6fadb-10bd-49e3-a778-290da82f7a8d               Facebook    1   0   
00ebf4a8-060f-4b99-93ac-c62724399483          Chrome Mobile    0   1   
...                                                     ...  ...  ..   
ffa08ff9-a132-4051-aef5-01a9c79367bc          Chrome Mobile    1   0   
ffb176df-ecd2-45d3-b05f-05b173a093a7          Chrome Mobile    1   0   
ffb79718-6f25-4896-b6b3-e58b80a6e147          Chrome Mobile    0   1   
ffca1153-c182-4f32-9e90-2a6008417497          Chrome Mobile    0   1   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b          Chrome Mobile    0   1   

                                      response day_of_the_week  responses  
auction_id                                                                 
0016d14a-ae18-4a02-a204-6ba53b52f2ed       1.0          Sunday        0.0  
008aafdf-deef-4482-8fec-d98e3da054da       1.0        Saturday        1.0  
00a1384a-5118-4d1b-925b-6cdada50318d       1.0          Monday        0.0  
00b6fadb-10bd-49e3-a778-290da82f7a8d       1.0       Wednesday        1.0  
00ebf4a8-060f-4b99-93ac-c62724399483       1.0          Friday        0.0  
...                                        ...             ...        ...  
ffa08ff9-a132-4051-aef5-01a9c79367bc       1.0          Sunday        1.0  
ffb176df-ecd2-45d3-b05f-05b173a093a7       1.0        Saturday        1.0  
ffb79718-6f25-4896-b6b3-e58b80a6e147       1.0        Thursday        0.0  
ffca1153-c182-4f32-9e90-2a6008417497       1.0          Friday        0.0  
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b       1.0          Friday        0.0  

[1243 rows x 11 columns]

In [5]:
#drop response column
df=df.drop(columns='response')
df

experiment        date  hour  \
auction_id                                                          
0016d14a-ae18-4a02-a204-6ba53b52f2ed    exposed  2020-07-05     2   
008aafdf-deef-4482-8fec-d98e3da054da    exposed  2020-07-04    16   
00a1384a-5118-4d1b-925b-6cdada50318d    exposed  2020-07-06     8   
00b6fadb-10bd-49e3-a778-290da82f7a8d    control  2020-07-08     4   
00ebf4a8-060f-4b99-93ac-c62724399483    control  2020-07-03    15   
...                                         ...         ...   ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc    exposed  2020-07-05    21   
ffb176df-ecd2-45d3-b05f-05b173a093a7    exposed  2020-07-04     1   
ffb79718-6f25-4896-b6b3-e58b80a6e147    control  2020-07-09     7   
ffca1153-c182-4f32-9e90-2a6008417497    control  2020-07-10    16   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b    exposed  2020-07-03     4   

                                             device_make  platform_os  \
auction_id                                                              
0016d14a-ae18-4a02-a204-6ba53b52f2ed               E5823            6   
008aafdf-deef-4482-8fec-d98e3da054da  Generic Smartphone            6   
00a1384a-5118-4d1b-925b-6cdada50318d  Generic Smartphone            6   
00b6fadb-10bd-49e3-a778-290da82f7a8d    Samsung SM-A202F            6   
00ebf4a8-060f-4b99-93ac-c62724399483  Generic Smartphone            6   
...                                                  ...          ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc  Generic Smartphone            6   
ffb176df-ecd2-45d3-b05f-05b173a093a7  Generic Smartphone            6   
ffb79718-6f25-4896-b6b3-e58b80a6e147  Generic Smartphone            6   
ffca1153-c182-4f32-9e90-2a6008417497  Generic Smartphone            6   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b  Generic Smartphone            6   

                                                    browser  yes  no  \
auction_id                                                             
0016d14a-ae18-4a02-a204-6ba53b52f2ed  Chrome Mobile WebView    0   1   
008aafdf-deef-4482-8fec-d98e3da054da          Chrome Mobile    1   0   
00a1384a-5118-4d1b-925b-6cdada50318d          Chrome Mobile    0   1   
00b6fadb-10bd-49e3-a778-290da82f7a8d               Facebook    1   0   
00ebf4a8-060f-4b99-93ac-c62724399483          Chrome Mobile    0   1   
...                                                     ...  ...  ..   
ffa08ff9-a132-4051-aef5-01a9c79367bc          Chrome Mobile    1   0   
ffb176df-ecd2-45d3-b05f-05b173a093a7          Chrome Mobile    1   0   
ffb79718-6f25-4896-b6b3-e58b80a6e147          Chrome Mobile    0   1   
ffca1153-c182-4f32-9e90-2a6008417497          Chrome Mobile    0   1   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b          Chrome Mobile    0   1   

                                     day_of_the_week  responses  
auction_id                                                       
0016d14a-ae18-4a02-a204-6ba53b52f2ed          Sunday        0.0  
008aafdf-deef-4482-8fec-d98e3da054da        Saturday        1.0  
00a1384a-5118-4d1b-925b-6cdada50318d          Monday        0.0  
00b6fadb-10bd-49e3-a778-290da82f7a8d       Wednesday        1.0  
00ebf4a8-060f-4b99-93ac-c62724399483          Friday        0.0  
...                                              ...        ...  
ffa08ff9-a132-4051-aef5-01a9c79367bc          Sunday        1.0  
ffb176df-ecd2-45d3-b05f-05b173a093a7        Saturday        1.0  
ffb79718-6f25-4896-b6b3-e58b80a6e147        Thursday        0.0  
ffca1153-c182-4f32-9e90-2a6008417497          Friday        0.0  
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b          Friday        0.0  

[1243 rows x 10 columns]

In [7]:
#drop browser column
new_df=df.drop(columns='browser')
new_df

experiment        date  hour  \
auction_id                                                          
0016d14a-ae18-4a02-a204-6ba53b52f2ed    exposed  2020-07-05     2   
008aafdf-deef-4482-8fec-d98e3da054da    exposed  2020-07-04    16   
00a1384a-5118-4d1b-925b-6cdada50318d    exposed  2020-07-06     8   
00b6fadb-10bd-49e3-a778-290da82f7a8d    control  2020-07-08     4   
00ebf4a8-060f-4b99-93ac-c62724399483    control  2020-07-03    15   
...                                         ...         ...   ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc    exposed  2020-07-05    21   
ffb176df-ecd2-45d3-b05f-05b173a093a7    exposed  2020-07-04     1   
ffb79718-6f25-4896-b6b3-e58b80a6e147    control  2020-07-09     7   
ffca1153-c182-4f32-9e90-2a6008417497    control  2020-07-10    16   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b    exposed  2020-07-03     4   

                                             device_make  platform_os  yes  \
auction_id                                                                   
0016d14a-ae18-4a02-a204-6ba53b52f2ed               E5823            6    0   
008aafdf-deef-4482-8fec-d98e3da054da  Generic Smartphone            6    1   
00a1384a-5118-4d1b-925b-6cdada50318d  Generic Smartphone            6    0   
00b6fadb-10bd-49e3-a778-290da82f7a8d    Samsung SM-A202F            6    1   
00ebf4a8-060f-4b99-93ac-c62724399483  Generic Smartphone            6    0   
...                                                  ...          ...  ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc  Generic Smartphone            6    1   
ffb176df-ecd2-45d3-b05f-05b173a093a7  Generic Smartphone            6    1   
ffb79718-6f25-4896-b6b3-e58b80a6e147  Generic Smartphone            6    0   
ffca1153-c182-4f32-9e90-2a6008417497  Generic Smartphone            6    0   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b  Generic Smartphone            6    0   

                                      no day_of_the_week  responses  
auction_id                                                           
0016d14a-ae18-4a02-a204-6ba53b52f2ed   1          Sunday        0.0  
008aafdf-deef-4482-8fec-d98e3da054da   0        Saturday        1.0  
00a1384a-5118-4d1b-925b-6cdada50318d   1          Monday        0.0  
00b6fadb-10bd-49e3-a778-290da82f7a8d   0       Wednesday        1.0  
00ebf4a8-060f-4b99-93ac-c62724399483   1          Friday        0.0  
...                                   ..             ...        ...  
ffa08ff9-a132-4051-aef5-01a9c79367bc   0          Sunday        1.0  
ffb176df-ecd2-45d3-b05f-05b173a093a7   0        Saturday        1.0  
ffb79718-6f25-4896-b6b3-e58b80a6e147   1        Thursday        0.0  
ffca1153-c182-4f32-9e90-2a6008417497   1          Friday        0.0  
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b   1          Friday        0.0  

[1243 rows x 9 columns]

In [8]:
#drop date column
new_df=new_df.drop(columns='date')
new_df

experiment  hour         device_make  \
auction_id                                                                  
0016d14a-ae18-4a02-a204-6ba53b52f2ed    exposed     2               E5823   
008aafdf-deef-4482-8fec-d98e3da054da    exposed    16  Generic Smartphone   
00a1384a-5118-4d1b-925b-6cdada50318d    exposed     8  Generic Smartphone   
00b6fadb-10bd-49e3-a778-290da82f7a8d    control     4    Samsung SM-A202F   
00ebf4a8-060f-4b99-93ac-c62724399483    control    15  Generic Smartphone   
...                                         ...   ...                 ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc    exposed    21  Generic Smartphone   
ffb176df-ecd2-45d3-b05f-05b173a093a7    exposed     1  Generic Smartphone   
ffb79718-6f25-4896-b6b3-e58b80a6e147    control     7  Generic Smartphone   
ffca1153-c182-4f32-9e90-2a6008417497    control    16  Generic Smartphone   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b    exposed     4  Generic Smartphone   

                                      platform_os  yes  no day_of_the_week  \
auction_id                                                                   
0016d14a-ae18-4a02-a204-6ba53b52f2ed            6    0   1          Sunday   
008aafdf-deef-4482-8fec-d98e3da054da            6    1   0        Saturday   
00a1384a-5118-4d1b-925b-6cdada50318d            6    0   1          Monday   
00b6fadb-10bd-49e3-a778-290da82f7a8d            6    1   0       Wednesday   
00ebf4a8-060f-4b99-93ac-c62724399483            6    0   1          Friday   
...                                           ...  ...  ..             ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc            6    1   0          Sunday   
ffb176df-ecd2-45d3-b05f-05b173a093a7            6    1   0        Saturday   
ffb79718-6f25-4896-b6b3-e58b80a6e147            6    0   1        Thursday   
ffca1153-c182-4f32-9e90-2a6008417497            6    0   1          Friday   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b            6    0   1          Friday   

                                      responses  
auction_id                                       
0016d14a-ae18-4a02-a204-6ba53b52f2ed        0.0  
008aafdf-deef-4482-8fec-d98e3da054da        1.0  
00a1384a-5118-4d1b-925b-6cdada50318d        0.0  
00b6fadb-10bd-49e3-a778-290da82f7a8d        1.0  
00ebf4a8-060f-4b99-93ac-c62724399483        0.0  
...                                         ...  
ffa08ff9-a132-4051-aef5-01a9c79367bc        1.0  
ffb176df-ecd2-45d3-b05f-05b173a093a7        1.0  
ffb79718-6f25-4896-b6b3-e58b80a6e147        0.0  
ffca1153-c182-4f32-9e90-2a6008417497        0.0  
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b        0.0  

[1243 rows x 8 columns]

In [9]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1243 entries, 0016d14a-ae18-4a02-a204-6ba53b52f2ed to ffdfdc09-48c7-4bfb-80f8-ec1eb633602b
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   experiment       1243 non-null   object 
 1   hour             1243 non-null   int64  
 2   device_make      1243 non-null   object 
 3   platform_os      1243 non-null   int64  
 4   yes              1243 non-null   int64  
 5   no               1243 non-null   int64  
 6   day_of_the_week  1243 non-null   object 
 7   responses        1243 non-null   float64
dtypes: float64(1), int64(4), object(3)
memory usage: 87.4+ KB


In [10]:
new_df.columns

Index(['experiment', 'hour', 'device_make', 'platform_os', 'yes', 'no',
       'day_of_the_week', 'responses'],
      dtype='object')

In [11]:
#numerical variables
cols_num= ['hour','platform_os', 'yes', 'no']
#categorical variables
cols_cat=['experiment','device_make','day_of_the_week']

In [12]:
df[cols_cat]

experiment         device_make  \
auction_id                                                            
0016d14a-ae18-4a02-a204-6ba53b52f2ed    exposed               E5823   
008aafdf-deef-4482-8fec-d98e3da054da    exposed  Generic Smartphone   
00a1384a-5118-4d1b-925b-6cdada50318d    exposed  Generic Smartphone   
00b6fadb-10bd-49e3-a778-290da82f7a8d    control    Samsung SM-A202F   
00ebf4a8-060f-4b99-93ac-c62724399483    control  Generic Smartphone   
...                                         ...                 ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc    exposed  Generic Smartphone   
ffb176df-ecd2-45d3-b05f-05b173a093a7    exposed  Generic Smartphone   
ffb79718-6f25-4896-b6b3-e58b80a6e147    control  Generic Smartphone   
ffca1153-c182-4f32-9e90-2a6008417497    control  Generic Smartphone   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b    exposed  Generic Smartphone   

                                     day_of_the_week  
auction_id                                            
0016d14a-ae18-4a02-a204-6ba53b52f2ed          Sunday  
008aafdf-deef-4482-8fec-d98e3da054da        Saturday  
00a1384a-5118-4d1b-925b-6cdada50318d          Monday  
00b6fadb-10bd-49e3-a778-290da82f7a8d       Wednesday  
00ebf4a8-060f-4b99-93ac-c62724399483          Friday  
...                                              ...  
ffa08ff9-a132-4051-aef5-01a9c79367bc          Sunday  
ffb176df-ecd2-45d3-b05f-05b173a093a7        Saturday  
ffb79718-6f25-4896-b6b3-e58b80a6e147        Thursday  
ffca1153-c182-4f32-9e90-2a6008417497          Friday  
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b          Friday  

[1243 rows x 3 columns]

In [13]:
#one-hot encoding
cols_new_cat=pd.get_dummies(df[cols_cat],drop_first=True)
cols_new_cat

experiment_exposed  device_make_ANE-LX1  \
auction_id                                                                      
0016d14a-ae18-4a02-a204-6ba53b52f2ed                   1                    0   
008aafdf-deef-4482-8fec-d98e3da054da                   1                    0   
00a1384a-5118-4d1b-925b-6cdada50318d                   1                    0   
00b6fadb-10bd-49e3-a778-290da82f7a8d                   0                    0   
00ebf4a8-060f-4b99-93ac-c62724399483                   0                    0   
...                                                  ...                  ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc                   1                    0   
ffb176df-ecd2-45d3-b05f-05b173a093a7                   1                    0   
ffb79718-6f25-4896-b6b3-e58b80a6e147                   0                    0   
ffca1153-c182-4f32-9e90-2a6008417497                   0                    0   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b                   1                    0   

                                      device_make_CLT-L09  \
auction_id                                                  
0016d14a-ae18-4a02-a204-6ba53b52f2ed                    0   
008aafdf-deef-4482-8fec-d98e3da054da                    0   
00a1384a-5118-4d1b-925b-6cdada50318d                    0   
00b6fadb-10bd-49e3-a778-290da82f7a8d                    0   
00ebf4a8-060f-4b99-93ac-c62724399483                    0   
...                                                   ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc                    0   
ffb176df-ecd2-45d3-b05f-05b173a093a7                    0   
ffb79718-6f25-4896-b6b3-e58b80a6e147                    0   
ffca1153-c182-4f32-9e90-2a6008417497                    0   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b                    0   

                                      device_make_DLI-L22  device_make_E5823  \
auction_id                                                                     
0016d14a-ae18-4a02-a204-6ba53b52f2ed                    0                  1   
008aafdf-deef-4482-8fec-d98e3da054da                    0                  0   
00a1384a-5118-4d1b-925b-6cdada50318d                    0                  0   
00b6fadb-10bd-49e3-a778-290da82f7a8d                    0                  0   
00ebf4a8-060f-4b99-93ac-c62724399483                    0                  0   
...                                                   ...                ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc                    0                  0   
ffb176df-ecd2-45d3-b05f-05b173a093a7                    0                  0   
ffb79718-6f25-4896-b6b3-e58b80a6e147                    0                  0   
ffca1153-c182-4f32-9e90-2a6008417497                    0                  0   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b                    0                  0   

                                      device_make_ELE-L09  \
auction_id                                                  
0016d14a-ae18-4a02-a204-6ba53b52f2ed                    0   
008aafdf-deef-4482-8fec-d98e3da054da                    0   
00a1384a-5118-4d1b-925b-6cdada50318d                    0   
00b6fadb-10bd-49e3-a778-290da82f7a8d                    0   
00ebf4a8-060f-4b99-93ac-c62724399483                    0   
...                                                   ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc                    0   
ffb176df-ecd2-45d3-b05f-05b173a093a7                    0   
ffb79718-6f25-4896-b6b3-e58b80a6e147                    0   
ffca1153-c182-4f32-9e90-2a6008417497                    0   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b                    0   

                                      device_make_ELE-L29  \
auction_id                                                  
0016d14a-ae18-4a02-a204-6ba53b52f2ed                    0   
008aafdf-deef-4482-8fec-d98e3da054da                    0   
00a1384a-5118-4d1b-925b-6cdada50318d                    0   
00b6fadb-10bd-49e3-a778-290da82f7a8d              

In [14]:
#add the one-hot encoded columns to the dataframe
new_df=pd.concat([new_df,cols_new_cat],axis=1)
new_df

experiment  hour         device_make  \
auction_id                                                                  
0016d14a-ae18-4a02-a204-6ba53b52f2ed    exposed     2               E5823   
008aafdf-deef-4482-8fec-d98e3da054da    exposed    16  Generic Smartphone   
00a1384a-5118-4d1b-925b-6cdada50318d    exposed     8  Generic Smartphone   
00b6fadb-10bd-49e3-a778-290da82f7a8d    control     4    Samsung SM-A202F   
00ebf4a8-060f-4b99-93ac-c62724399483    control    15  Generic Smartphone   
...                                         ...   ...                 ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc    exposed    21  Generic Smartphone   
ffb176df-ecd2-45d3-b05f-05b173a093a7    exposed     1  Generic Smartphone   
ffb79718-6f25-4896-b6b3-e58b80a6e147    control     7  Generic Smartphone   
ffca1153-c182-4f32-9e90-2a6008417497    control    16  Generic Smartphone   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b    exposed     4  Generic Smartphone   

                                      platform_os  yes  no day_of_the_week  \
auction_id                                                                   
0016d14a-ae18-4a02-a204-6ba53b52f2ed            6    0   1          Sunday   
008aafdf-deef-4482-8fec-d98e3da054da            6    1   0        Saturday   
00a1384a-5118-4d1b-925b-6cdada50318d            6    0   1          Monday   
00b6fadb-10bd-49e3-a778-290da82f7a8d            6    1   0       Wednesday   
00ebf4a8-060f-4b99-93ac-c62724399483            6    0   1          Friday   
...                                           ...  ...  ..             ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc            6    1   0          Sunday   
ffb176df-ecd2-45d3-b05f-05b173a093a7            6    1   0        Saturday   
ffb79718-6f25-4896-b6b3-e58b80a6e147            6    0   1        Thursday   
ffca1153-c182-4f32-9e90-2a6008417497            6    0   1          Friday   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b            6    0   1          Friday   

                                      responses  experiment_exposed  \
auction_id                                                            
0016d14a-ae18-4a02-a204-6ba53b52f2ed        0.0                   1   
008aafdf-deef-4482-8fec-d98e3da054da        1.0                   1   
00a1384a-5118-4d1b-925b-6cdada50318d        0.0                   1   
00b6fadb-10bd-49e3-a778-290da82f7a8d        1.0                   0   
00ebf4a8-060f-4b99-93ac-c62724399483        0.0                   0   
...                                         ...                 ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc        1.0                   1   
ffb176df-ecd2-45d3-b05f-05b173a093a7        1.0                   1   
ffb79718-6f25-4896-b6b3-e58b80a6e147        0.0                   0   
ffca1153-c182-4f32-9e90-2a6008417497        0.0                   0   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b        0.0                   1   

                                      device_make_ANE-LX1  ...  \
auction_id                                                 ...   
0016d14a-ae18-4a02-a204-6ba53b52f2ed                    0  ...   
008aafdf-deef-4482-8fec-d98e3da054da                    0  ...   
00a1384a-5118-4d1b-925b-6cdada50318d                    0  ...   
00b6fadb-10bd-49e3-a778-290da82f7a8d                    0  ...   
00ebf4a8-060f-4b99-93ac-c62724399483                    0  ...   
...                                                   ...  ...   
ffa08ff9-a132-4051-aef5-01a9c79367bc                    0  ...   
ffb176df-ecd2-45d3-b05f-05b173a093a7                    0  ...   
ffb79718-6f25-4896-b6b3-e58b80a6e147                    0  ...   
ffca1153-c182-4f32-9e90-2a6008417497                    0  ...   
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b                    0  ...   

                                      device_make_XiaoMi Redmi Note 7  \
auction_id                                                              
0016d14a-ae18-4a02-a204-6ba53b52f2ed                                0   
008aafdf-deef-4482

In [15]:
all_cols_cat=list(cols_new_cat.columns)
all_cols_cat

['experiment_exposed',
 'device_make_ANE-LX1',
 'device_make_CLT-L09',
 'device_make_DLI-L22',
 'device_make_E5823',
 'device_make_ELE-L09',
 'device_make_ELE-L29',
 'device_make_EML-L09',
 'device_make_EVA-L09',
 'device_make_EVR-N29',
 'device_make_F3311',
 'device_make_FIG-LX1',
 'device_make_G3311',
 'device_make_Generic Smartphone',
 'device_make_HTC 10',
 'device_make_HTC Desire $2',
 'device_make_HTC One $2',
 'device_make_HTC U12+',
 'device_make_Huawei Browser',
 'device_make_I3312',
 'device_make_IMO $2',
 'device_make_LG-$2',
 'device_make_LYA-L09',
 'device_make_MAR-LX1A',
 'device_make_MRD-LX1',
 'device_make_Moto $2',
 'device_make_Nokia undefined$2$3',
 'device_make_Nokia$2$3',
 'device_make_OnePlus ONEPLUS A3000',
 'device_make_OnePlus ONEPLUS A5000',
 'device_make_OnePlus ONEPLUS A5010',
 'device_make_OnePlus ONEPLUS A6013',
 'device_make_PCT-L29',
 'device_make_POT-LX1',
 'device_make_PRA-LX1',
 'device_make_Pixel 2',
 'device_make_Pixel 3',
 'device_make_Pixel 3a',
 

In [36]:
X=new_df[all_cols_cat]
X.shape


(1243, 121)

In [18]:
cols_new_num=['hour','platform_os', 'yes', 'no']
new_df[cols_new_num]

hour  platform_os  yes  no
auction_id                                                      
0016d14a-ae18-4a02-a204-6ba53b52f2ed     2            6    0   1
008aafdf-deef-4482-8fec-d98e3da054da    16            6    1   0
00a1384a-5118-4d1b-925b-6cdada50318d     8            6    0   1
00b6fadb-10bd-49e3-a778-290da82f7a8d     4            6    1   0
00ebf4a8-060f-4b99-93ac-c62724399483    15            6    0   1
...                                    ...          ...  ...  ..
ffa08ff9-a132-4051-aef5-01a9c79367bc    21            6    1   0
ffb176df-ecd2-45d3-b05f-05b173a093a7     1            6    1   0
ffb79718-6f25-4896-b6b3-e58b80a6e147     7            6    0   1
ffca1153-c182-4f32-9e90-2a6008417497    16            6    0   1
ffdfdc09-48c7-4bfb-80f8-ec1eb633602b     4            6    0   1

[1243 rows x 4 columns]

In [19]:
cols_input=cols_new_num+all_cols_cat

cols_input

['hour',
 'platform_os',
 'yes',
 'no',
 'experiment_exposed',
 'device_make_ANE-LX1',
 'device_make_CLT-L09',
 'device_make_DLI-L22',
 'device_make_E5823',
 'device_make_ELE-L09',
 'device_make_ELE-L29',
 'device_make_EML-L09',
 'device_make_EVA-L09',
 'device_make_EVR-N29',
 'device_make_F3311',
 'device_make_FIG-LX1',
 'device_make_G3311',
 'device_make_Generic Smartphone',
 'device_make_HTC 10',
 'device_make_HTC Desire $2',
 'device_make_HTC One $2',
 'device_make_HTC U12+',
 'device_make_Huawei Browser',
 'device_make_I3312',
 'device_make_IMO $2',
 'device_make_LG-$2',
 'device_make_LYA-L09',
 'device_make_MAR-LX1A',
 'device_make_MRD-LX1',
 'device_make_Moto $2',
 'device_make_Nokia undefined$2$3',
 'device_make_Nokia$2$3',
 'device_make_OnePlus ONEPLUS A3000',
 'device_make_OnePlus ONEPLUS A5000',
 'device_make_OnePlus ONEPLUS A5010',
 'device_make_OnePlus ONEPLUS A6013',
 'device_make_PCT-L29',
 'device_make_POT-LX1',
 'device_make_PRA-LX1',
 'device_make_Pixel 2',
 'device_m

In [20]:
len(cols_input)

125

In [21]:
df_data=new_df[cols_input+['responses']]
df_data.head()

hour  platform_os  yes  no  \
auction_id                                                         
0016d14a-ae18-4a02-a204-6ba53b52f2ed     2            6    0   1   
008aafdf-deef-4482-8fec-d98e3da054da    16            6    1   0   
00a1384a-5118-4d1b-925b-6cdada50318d     8            6    0   1   
00b6fadb-10bd-49e3-a778-290da82f7a8d     4            6    1   0   
00ebf4a8-060f-4b99-93ac-c62724399483    15            6    0   1   

                                      experiment_exposed  device_make_ANE-LX1  \
auction_id                                                                      
0016d14a-ae18-4a02-a204-6ba53b52f2ed                   1                    0   
008aafdf-deef-4482-8fec-d98e3da054da                   1                    0   
00a1384a-5118-4d1b-925b-6cdada50318d                   1                    0   
00b6fadb-10bd-49e3-a778-290da82f7a8d                   0                    0   
00ebf4a8-060f-4b99-93ac-c62724399483                   0                    0   

                                      device_make_CLT-L09  \
auction_id                                                  
0016d14a-ae18-4a02-a204-6ba53b52f2ed                    0   
008aafdf-deef-4482-8fec-d98e3da054da                    0   
00a1384a-5118-4d1b-925b-6cdada50318d                    0   
00b6fadb-10bd-49e3-a778-290da82f7a8d                    0   
00ebf4a8-060f-4b99-93ac-c62724399483                    0   

                                      device_make_DLI-L22  device_make_E5823  \
auction_id                                                                     
0016d14a-ae18-4a02-a204-6ba53b52f2ed                    0                  1   
008aafdf-deef-4482-8fec-d98e3da054da                    0                  0   
00a1384a-5118-4d1b-925b-6cdada50318d                    0                  0   
00b6fadb-10bd-49e3-a778-290da82f7a8d                    0                  0   
00ebf4a8-060f-4b99-93ac-c62724399483                    0                  0   

                                      device_make_ELE-L09  ...  \
auction_id                                                 ...   
0016d14a-ae18-4a02-a204-6ba53b52f2ed                    0  ...   
008aafdf-deef-4482-8fec-d98e3da054da                    0  ...   
00a1384a-5118-4d1b-925b-6cdada50318d                    0  ...   
00b6fadb-10bd-49e3-a778-290da82f7a8d                    0  ...   
00ebf4a8-060f-4b99-93ac-c62724399483                    0  ...   

                                      device_make_XiaoMi Redmi S2  \
auction_id                                                          
0016d14a-ae18-4a02-a204-6ba53b52f2ed                            0   
008aafdf-deef-4482-8fec-d98e3da054da                            0   
00a1384a-5118-4d1b-925b-6cdada50318d                            0   
00b6fadb-10bd-49e3-a778-290da82f7a8d                            0   
00ebf4a8-060f-4b99-93ac-c62724399483                            0   

                                      device_make_YAL-L21  device_make_iPhone  \
auction_id                                                                      
0016d14a-ae18-4a02-a204-6ba53b52f2ed                    0                   0   
008aafdf-deef-4482-8fec-d98e3da054da                    0                   0   
00a1384a-5118-4d1b-925b-6cdada50318d                    0                   0   
00b6fadb-10bd-49e3-a778-290da82f7a8d                    0                   0   
00ebf4a8-060f-4b99-93ac-c62724399483                    0                   0   

                                      day_of_the_week_Monday  \
auction_id                                                     
0016d14a-ae18-4a02-a204-6ba53b52f2ed                       0   
008aafdf-deef-4482-8fec-d98e3da054da                       0   
00a1384a-5118-4d1b-925b-6cdada50318d                       1   
00b6fadb-10bd-49e3-a778-290da82f7a8d                       0   
00ebf4a8-060f-4b99-93ac-c62724399483                       0   

                                      da

In [22]:
#shuffle the samples

df_data=df_data.sample(n=len(df_data))
df_data=df_data.reset_index(drop=True)

In [23]:
#30% of validation and test samples
df_valid_test=df_data.sample(frac=0.3)
#splits test and validation samples by 2/3
df_test=df_valid_test.sample(frac=2/3) 
df_validation=df_valid_test.drop(df_test.index)


In [24]:
#training sample
df_train=df_data.drop(df_valid_test.index)

# Decision Tree


In [53]:
# Split dataset into training set and test set
from sklearn.model_selection import train_test_split # Import train_test_split function
from sklearn import metrics #Import scikit-learn metrics module for accuracy calculation
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix

X=new_df[all_cols_cat]
y=df_data['responses']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0) # 70% training and 30% test
# Create Decision Tree classifer object
clf = DecisionTreeClassifier()

# Train Decision Tree Classifer
clf = clf.fit(X_train,y_train)

#Predict the response for test dataset
y_pred = clf.predict(X_test)

# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))


Accuracy: 0.5656836461126006


In [54]:
#importance of the features
feature_importance=pd.DataFrame(tree.feature_importances_ ,index=cols_input,columns=['Importance']).sort_values('Importance',ascending=False)
feature_importance

Importance
yes                                       1.0
hour                                      0.0
device_make_Samsung SM-G980F              0.0
device_make_Samsung SM-J610FN             0.0
device_make_Samsung SM-J600FN             0.0
...                                       ...
device_make_POT-LX1                       0.0
device_make_PCT-L29                       0.0
device_make_OnePlus ONEPLUS A6013         0.0
device_make_OnePlus ONEPLUS A5010         0.0
day_of_the_week_Wednesday                 0.0

[125 rows x 1 columns]

# Logistic Regression

In [55]:
from sklearn.linear_model import LogisticRegression
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
logreg = LogisticRegression()
logreg.fit(X_train, y_train)

LogisticRegression()

In [56]:
y_pred = logreg.predict(X_test)
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(logreg.score(X_test, y_test)))

Accuracy of logistic regression classifier on test set: 0.52


In [57]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_pred)
print(confusion_matrix)

[[139  67]
 [111  56]]


This means the model has (139+56) correct predictions and (11+67)incorrect predictions

In [52]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         0.0       0.56      0.67      0.61       206
         1.0       0.46      0.34      0.39       167

    accuracy                           0.52       373
   macro avg       0.51      0.51      0.50       373
weighted avg       0.51      0.52      0.51       373



In [ ]:
#feature importance


# XGBoost

In [58]:
import xgboost as xgb

ModuleNotFoundError: No module named 'xgboost'

In [ ]:
pip install xgboost 